infrastructure/src/main/java/org/corfudb/infrastructure/remotecorfutable/DatabaseHandler.java

```diff
@@ -0,0 +1,472 @@
+package org.corfudb.infrastructure.remotecorfutable;
+
+import com.google.errorprone.annotations.DoNotCall;
+import com.google.protobuf.ByteString;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.checkerframework.checker.index.qual.Positive;
+import org.corfudb.infrastructure.remotecorfutable.utils.KeyEncodingUtil;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuError;
+import org.rocksdb.BuiltinComparator;
+import org.rocksdb.ColumnFamilyDescriptor;
+import org.rocksdb.ColumnFamilyHandle;
+import org.rocksdb.ColumnFamilyOptions;
+import org.rocksdb.Options;
+import org.rocksdb.ReadOptions;
+import org.rocksdb.RocksDB;
+import org.rocksdb.RocksDBException;
+import org.rocksdb.RocksIterator;
+
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ThreadPoolExecutor;
+
+import static org.corfudb.infrastructure.remotecorfutable.utils.DatabaseConstants.LATEST_VERSION_READ;
+import static org.corfudb.infrastructure.remotecorfutable.utils.DatabaseConstants.DATABASE_CHARSET;
+import static org.corfudb.infrastructure.remotecorfutable.utils.DatabaseConstants.METADATA_COLUMN_CACHE_SIZE;
+import static org.corfudb.infrastructure.remotecorfutable.utils.DatabaseConstants.METADATA_COLUMN_SUFFIX;
+
+/**
+ * The DatabaseHandler provides an interface for the RocksDB instance storing data for server side
+ * RemoteCorfuTables.
+ *
+ * <p>Created by nvaishampayan517 on 7/23/21.
+ */
+@Slf4j
+public class DatabaseHandler implements AutoCloseable {
```

Comment:
There is [PersistedStreamingMap](https://github.com/CorfuDB/CorfuDB/blob/d5443d431eb3f11a54dfebd95c19473f9cae23d6/runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java#L38) which contains almost the same code, so we have code duplication. Is it possible to create another PR right now and extract common code from DatabaseHandler and PersistedStreamingMap?

We'd have another smaller PR, faster to review, less probability to have a bug. 
If that would work for you then would be great to have another pr.

Reply:
Then this PR will use that smaller pr

Reply:
From what I see, only the outward facing API is the same between the persisted streaming map and this handler. In this handler, we're enforcing strict Key formats as well as supporting the new record-level MVCC implementation. So under the hood, the methods are doing things like getting recent versions through iteration, where in the PersistentStreamingMap the same method would be implemented using just GET calls.